keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -469,19 +519,269 @@ states (e.g. not yet block resources) will help with improving the scheduling ac
-
 Addressing requirement (8) is the biggest effort as it requires much closer integration between
 scheduler and autoscaling components. So in the initial steps we will only focus on mitigating
 this problem with existing mechanisms (e.g. reserving resources via NominatedNodeName).
 
 However, approval for this KEP is NOT an approval for this vision. We only sketch it to show that
 we see a viable path forward from the proposed design that will not require significant rework.
 
+### Scheduler Changes for Beta
+
+For the `Alpha` phase, we focused on plumbing the `Workload` API and implementing
+the `GangScheduling` plugin using simple barriers (`PreEnqueue` and `Permit`).
+While this satisfied the correctness requirement for "all-or-nothing" scheduling,
+it did not address performance or efficiency at scale, scheduling deadlocks,
+nor did it solve the problem of partial preemption application.
+
+For `Beta`, we propose introducing a **Workload Scheduling Cycle**.
+This mechanism processes all Pods belonging to a single `PodGroup` in one batch,
+rather than attempting to schedule them individually in isolation using the
+traditional pod-by-pod approach.
+While this won't fully address the "optimal enough" part of requirement (2),
+it ensures that all gang pods are processed together.
+The single scheduling cycle, together with blocking resources using nomination,
+will address requirement (3).
+
+We will also introduce delayed preemption (described in [KEP-5710](https://kep.k8s.io/5711)).
+Together with the introduction of a dedicated Workload Scheduling Cycle,
+this will address requirement (5).
+
+#### The Workload Scheduling Cycle
+
+We introduce a new phase in the main scheduling loop (`scheduleOne`). In the
+end-to-end Pod scheduling flow, it is planned to place this new phase *before*
+the standard pod-by-pod scheduling cycle.
+
+When the scheduler pops a Pod from the active queue, it checks if that Pod
+belongs to an unscheduled `PodGroup` with a `Gang` policy. If so, the scheduler
+initiates the Workload Scheduling Cycle.
+
+```md
+<<[UNRESOLVED Scope of the Cycle]>>
+It is currently unresolved whether the Workload Scheduling Cycle should operate
+on the entire `Workload` object (handling all defined PodGroups simultaneously)
+or strictly at the `PodGroup` level.
+
+* PodGroup Level: The cycle processes only the specific `PodGroup` (and replica key)
+  associated with the popped Pod. This is simpler and aligns with
+  the Gang Scheduling definition and current implementation.
+* Workload Level: The cycle attempts to schedule all PodGroups within the Workload.
+  This allows for complex dependencies between groups but increases the complexity
+  and doesn't bring immediate added value.
+
+*Proposed:* Implement it on PodGroup Level for Beta. However, future migration
+to the Workload Level might necessitate non-trivial changes to the phase
+introduced by this KEP.
```

Comment:
+1 to this - let's start with PodGroup replica cycle

I agree that it may require non-trivial changes in the future, but I also claim that these will be algorithm-oriented and localized to this single place. Which imho is the exact situation where we want to make such tradeoffs for the sake of making progress incrementally.

Reply:
+1

Complex true workload controllers should implement any orchestration themselves. In other words, if there really are, say, dependencies between PodGroups, I think we should leave that complexity in the upper layer controller, and the scheduler just deal in PodGroups. So the controller would wait to create the second PodGroup until after it made sure the first one got scheduled. This does mean we don't have full workload atomic scheduling, only PodGroup atomic scheduling. But that level of complexity probably needs something more like Reservation.